cou/utils/juju_utils.py

```diff
@@ -396,3 +397,107 @@ async def async_upgrade_charm(
+        """Own timeout exception."""
+
+    def __init__(self, model: Model):
+        """Initialize.
+
+        :param model: model to wait
+        :type model: Model
+        """
+        self.model = model
+        self.model_name = self.model.info.name
+        self.timeout = timedelta(seconds=JujuWaiter.DEFAULT_TIMEOUT)
+        self.start_time = datetime.now()
+        self.log = logging.getLogger(self.__class__.__name__)
+
+    async def wait(self, timeout_seconds: int = DEFAULT_TIMEOUT) -> None:
+        """Wait model to stabilize.
+
+        :param timeout_seconds: wait model till timeout_seconds. If passed raise TimeoutException
+        :type timeout_seconds: Optional[int]
+        :return: None
+        :raises TimeoutException: if timeout_seconds passed.
+        :raises JujuMachineError: if it has a machine exception
+        :raises JujuAgentError: if it has an agent exception
+        :raises JujuUnitError: if it has unit exception
+        :raises JujuAppError: if it has application exception
+        """
+        self.start_time = datetime.now()
+        if timeout_seconds:
+            self.timeout = timedelta(seconds=timeout_seconds)
+        self.log.debug("Waiting to stabilize in %s seconds", self.timeout)
+
+        while True:
+            await self._ensure_model_connected()
+            try:
+                await self.model.wait_for_idle(
+                    idle_period=JujuWaiter.MODEL_IDLE_PERIOD,
+                    timeout=JujuWaiter.JUJU_IDLE_TIMEOUT,
+                )
+                self.log.debug(
+                    "Model %s is idle for %s seconds.",
+                    self.model.info.name,
+                    JujuWaiter.MODEL_IDLE_PERIOD,
+                )
+                return
+            except (
+                JujuMachineError,
+                JujuAgentError,
+                JujuUnitError,
+                JujuAppError,
+            ):
+                raise
+            except Exception as ex:
+                # We do not care exceptions other than Juju(Machine|Agent|Unit|App)Error because
+                # when juju connection is dropped you can have wide range of exceptions depending
+                # on the case
+                self.log.debug("Unknown error while waiting to stabilize", exc_info=ex)
+
+            self._check_time()
+
+    async def _ensure_model_connected(self) -> None:
+        """Ensure that the model is connected.
+
+        :raises JujuWaiter.TimeoutException: if timeout occurs
+        """
+        while _is_model_disconnected(self.model):
+            await _disconnect(self.model)
+            try:
+                self._check_time()
+                await self.model.connect_model(self.model_name)
+            except JujuWaiter.TimeoutException:
+                raise
+            except Exception:
+                self.log.debug(
+                    "Model has unexpected exception while connecting, retrying", exc_info=True
```

Comment:
I was wondering why we are suing `exc_info=True` here but at line 475 we use `exc_info=ex`? If I'm not mistaken, they have the same effect: attaching the exception instance to the log message. If that's true, could we unify the code for easier future maintenance?

Reply:
Correct version is `exc_info=True` [reference](https://docs.python.org/3.10/library/logging.html#logging.Logger.debug)
corrected